src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java

```diff
@@ -332,36 +299,53 @@ protected boolean isExecutable(PathFragment path) throws IOException {
     return m != null || super.isExecutable(path);
   }
 
+  @FunctionalInterface
+  interface PathOp {
+    void run(Path path) throws IOException;
+  }
+
   @Override
   protected void setReadable(PathFragment path, boolean readable) throws IOException {
-    RemoteFileArtifactValue m = getRemoteMetadata(path);
-    if (m == null) {
-      super.setReadable(path, readable);
+    doOnRemoteAndLocal(path, p -> p.setReadable(readable));
+  }
+
+  private void doOnRemoteAndLocal(PathFragment path, PathOp pathOp) throws IOException {
+    FileNotFoundException remoteException = null;
+    try {
+      pathOp.run(remoteOutputTree.getPath(path));
+    } catch (FileNotFoundException e) {
+      remoteException = e;
+    }
+
+
+    FileNotFoundException localException = null;
+    try {
+      pathOp.run(delegateFs.getPath(path));
+    } catch (FileNotFoundException e) {
+      localException = e;
     }
+
+    if (remoteException == null || localException == null) {
+      return;
+    }
+
+    localException.addSuppressed(remoteException);
+    throw localException;
   }
 
   @Override
   public void setWritable(PathFragment path, boolean writable) throws IOException {
-    RemoteFileArtifactValue m = getRemoteMetadata(path);
-    if (m == null) {
-      super.setWritable(path, writable);
-    }
+    doOnRemoteAndLocal(path, p -> p.setWritable(writable));
   }
 
   @Override
   protected void setExecutable(PathFragment path, boolean executable) throws IOException {
-    RemoteFileArtifactValue m = getRemoteMetadata(path);
-    if (m == null) {
-      super.setExecutable(path, executable);
-    }
+    doOnRemoteAndLocal(path, p -> p.setExecutable(executable));
   }
 
   @Override
   protected void chmod(PathFragment path, int mode) throws IOException {
-    RemoteFileArtifactValue m = getRemoteMetadata(path);
-    if (m == null) {
-      super.chmod(path, mode);
-    }
+    doOnRemoteAndLocal(path, p -> p.chmod(mode));
   }
```

Comment:
Do you mind adding some tests covering the various cases (both exist, none exist, only local, only remote)? This will make future regressions easier to catch since we won't have to debug them at  the Skyframe level :)

I think it's fine to only add tests for `chmod` and trust that the `set*` methods are correct, since they're structurally similar.

Reply:
My original plan was to extract the changes to RemoteActionFS to another CL and add tests there.

Reply:
Done with cc712ee5278a6b4686e2ac8fcec3bc54e086fe6e.